### PR TITLE
Gpu iewpf

### PR DESCRIPTION
**Pull request by havahol**
CUDA implementation of the IEWPF algorithm.

PR #146 broke some of the functionality, but after discussion with  @setmar, we have decided to fix this issue in the next PR.

The notebook `prototypes/IEWPF_GPU.ipynb` contains an example run of the GPU implementation, but the velocity field under the drifter is currently not correctly updated. Unit tests are found in `tests/particle_filter_tests.py 5`, and one of the tests is expected to fail.

Next task will be to add the interpolation operators used in the model error into the IEWPF algorithm, and the issues briefly mentioned above should be fixed in the process.

### CODE REVIEW

File: gpu_ocean/SWESimulators/IEWPFOcean.py [hunk 1]
```diff
@@ -0,0 +1,1238 @@
...
+        #print "-log(target_weight): ", -np.log(target_weight)
+        #print "exp(-target_weight): ", np.exp(-target_weight)
+        #print "1/Ne: ", 1.0/ensemble.getNumParticles()
+        #print "WWWWWWWWWWWWWWW"
+
+
+        for p in range(ensemble.getNumParticles()):
+            #iewpfOcean.debug = p==0
+
+            # Step 2: Sample xi Ìƒ N(0, P)
```
Comment (setmar) [inline gpu_ocean/SWESimulators/IEWPFOcean.py]: Weird character?

File: gpu_ocean/SWESimulators/IEWPFOcean.py [hunk 2]
```diff
@@ -0,0 +1,1238 @@
...
+                                        phi, xi, gamma, 
+                                        target_weight, w_rest[p])
+
+            eta_a = self._expand_to_periodic_boundaries(eta_a, 2)
+            hu_a  = self._expand_to_periodic_boundaries(hu_a,  2)
+            hv_a  = self._expand_to_periodic_boundaries(hv_a,  2)
+            ensemble.particles[p].upload(eta_a, hu_a, hv_a)
+
+            # TODO
+            #ensemble.particles[p].drifters.setDrifterPositions(newPos)
```
Comment (setmar) [inline gpu_ocean/SWESimulators/IEWPFOcean.py]: Inactive code should generally be removed.

File: gpu_ocean/SWESimulators/IEWPFOcean.py [hunk 3]
```diff
@@ -0,0 +1,1238 @@
...
+            return interior
+        (ny, nx) = interior.shape
+
+        nx_halo = nx + 2*ghostcells
+        ny_halo = ny + 2*ghostcells
+        newBuf = np.zeros((ny_halo, nx_halo))
+        newBuf[ghostcells:-ghostcells, ghostcells:-ghostcells] = interior 
+        for g in range(ghostcells):
+            newBuf[g, :] = newBuf[ny_halo - 2*ghostcells + g, :]
+            #newBuf[ny_halo - 2*ghostcells + g, :] *=0
```
Comment (setmar) [inline gpu_ocean/SWESimulators/IEWPFOcean.py]: Inactive code should generally be removed.

File: gpu_ocean/SWESimulators/IEWPFOcean.py [hunk 4]
```diff
@@ -0,0 +1,1238 @@
...
+        if ghostcells == 0:
+            return interior
+        (ny, nx) = interior.shape
+
+        nx_halo = nx + 2*ghostcells
+        ny_halo = ny + 2*ghostcells
+        newBuf = np.zeros((ny_halo, nx_halo))
+        newBuf[ghostcells:-ghostcells, ghostcells:-ghostcells] = interior 
+        for g in range(ghostcells):
+            newBuf[g, :] = newBuf[ny_halo - 2*ghostcells + g, :]
```
Comment (setmar) [inline gpu_ocean/SWESimulators/IEWPFOcean.py]: Comment offsets.

File: gpu_ocean/SWESimulators/IEWPFOcean.py [hunk 5]
```diff
@@ -0,0 +1,1238 @@
...
+            global_y_j = pos_y - 3 + loc_y_j
+            global_y_j = self._apply_periodic_boundary(global_y_j, self.ny)
+            for loc_x_j in range(7):
+                global_x_j = pos_x - 3 + loc_x_j
+                global_x_j = self._apply_periodic_boundary(global_x_j, self.nx)
+
+                global_j = global_y_j*self.nx + global_x_j
+                local_j = loc_y_j*7 + loc_x_j
+
+                #loc_vec[local_j] = glob_vec[global_j]
```
Comment (setmar) [inline gpu_ocean/SWESimulators/IEWPFOcean.py]: Inactive code should generally be removed.

File: gpu_ocean/SWESimulators/IEWPFOcean.py [hunk 6]
```diff
@@ -0,0 +1,1238 @@
...
+                                                                sim.small_scale_model_error.staggered, \
+                                                                sim.small_scale_model_error.soar_q0, \
+                                                                sim.small_scale_model_error.soar_L)
+            draw_from_q_maker.random_numbers_host = original.copy()
+            q_eta, q_hu, q_hv = draw_from_q_maker._obtainOceanPerturbations_CPU(H_mid, sim.f, sim.coriolis_beta, sim.g)
+            self.showMatrices(q_eta[1:-1, 1:-1], q_hu, "Equivalent sample from Q", q_hv)
+            self.showMatrices(p_eta[1:-1, 1:-1] - q_eta[1:-1, 1:-1], p_hu - q_hu, "diff sample from P and sample from Q", p_hv - q_hv)
+
+
+        #gamma_from_p = np.sum(p_eta[1:-1, 1:-1]**2) + np.sum(p_hu**2) + np.sum(p_hv**2)
```
Comment (setmar) [inline gpu_ocean/SWESimulators/IEWPFOcean.py]: Inactive code should generally be removed.

File: gpu_ocean/SWESimulators/IEWPFOcean.py [hunk 7]
```diff
@@ -0,0 +1,1238 @@
...
+            # 0.1) Find the cell index assuming no ghost cells
+            cell_id_x = int(np.floor(observed_drifter_position[0]/sim.dx))
+            cell_id_y = int(np.floor(observed_drifter_position[1]/sim.dy))
+            if self.debug: print ("(cell_id_x, cell_id_y): ", (cell_id_x, cell_id_y))
+
+            # 1) Solve linear problem
+            e = np.dot(self.S_host, local_innovation)
+            if self.debug: print("e: ", e)
+
+            # 2) K = QH^T e = U_GB Q^{1/2} Q^{1/2} U_GB^T  H^T e
```
Comment (setmar) [inline gpu_ocean/SWESimulators/IEWPFOcean.py]: Inactive code should generally be removed.

File: gpu_ocean/SWESimulators/IEWPFOcean.py [hunk 8]
```diff
@@ -0,0 +1,1238 @@
...
+            # 2.2) Apply U_GB Q^{1/2} to the result
+            # 2.2.1)  Easiest way: map local_eta to a global K_eta_tmp buffer
+            K_eta_tmp = np.zeros((sim.ny, sim.nx))
+            if self.debug: print( "K_eta_tmp.shape",  K_eta_tmp.shape)
+            for j in range(7):
+                j_global = (cell_id_y-3+j+sim.ny)%sim.ny 
+                for i in range(7):
+                    i_global = (cell_id_x-3+i+sim.nx)%sim.nx 
+                    K_eta_tmp[j_global, i_global] += local_eta[j,i]
+                    #K_eta_tmp[j_global, i_global] += 10000000*local_eta[j,i]
```
Comment (setmar) [inline gpu_ocean/SWESimulators/IEWPFOcean.py]: Inactive code should generally be removed.

File: gpu_ocean/SWESimulators/IEWPFOcean.py [hunk 9]
```diff
@@ -0,0 +1,1238 @@
...
+            # 2.2.3) Apply soar + geo-balance
+            H_mid = sim.downloadBathymetry()[0]
+            K_eta , K_hu, K_hv = sim.small_scale_model_error._obtainOceanPerturbations_CPU(H_mid, sim.f, sim.coriolis_beta, sim.g)
+            if self.debug: self.showMatrices(K_eta[1:-1, 1:-1], K_hu, "Kalman gain from drifter " + str(drifter), K_hv)
+
+            total_K_eta += K_eta[1:-1, 1:-1]
+            total_K_hu  += K_hu
+            total_K_hv  += K_hv
+            if self.debug: self.showMatrices(total_K_eta, total_K_hu, "Total Kalman gain after drifter " + str(drifter), total_K_hv)
+            #showMatrices(total_K_eta, total_K_hu, "Total Kalman gain after drifter " + str(drifter), total_K_hv)
```
Comment (setmar) [inline gpu_ocean/SWESimulators/IEWPFOcean.py]: Inactive code should generally be removed.

File: gpu_ocean/SWESimulators/IEWPFOcean.py [hunk 10]
```diff
@@ -0,0 +1,1238 @@
...
+        """
+        Solving the scalar implicit equation using the Lambert W function, and 
+        updating the buffers eta_a, hu_a, hv_a as:
+        x_a = x_a + alpha*xi
+        """
+
+
+        # 5) obtain gamma
+        if self.debug: print ("Shapes of xi: ", xi[0].shape, xi[1].shape, xi[2].shape)
+        #gamma = 0.0
```
Comment (setmar) [inline gpu_ocean/SWESimulators/IEWPFOcean.py]: Inactive code should generally be removed.

File: gpu_ocean/SWESimulators/IEWPFOcean.py [hunk 11]
```diff
@@ -0,0 +1,1238 @@
...
+        # 6) Find a
+        a = phi - w_rest + target_weight
+        if self.debug: print( "a = phi - w_rest + target_weight: ", a)
+            
+        c = phi - w_rest
+        if self.debug: 
+            print ("c = phi - w_rest: ", c)
+
+        # 7) Solving the Lambert W function
+        #alpha = 10000
```
Comment (setmar) [inline gpu_ocean/SWESimulators/IEWPFOcean.py]: Inactive code should generally be removed.

File: gpu_ocean/SWESimulators/IEWPFOcean.py [hunk 12]
```diff
@@ -0,0 +1,1238 @@
...
+            print ("Particle ID: ", particle_id)
+            print ("Obtained (alpha k=0, alpha k=-1): ", (alpha_zero, alpha_min1))
+            print ("The requirement is lamber_W_arg > (-1.0/exp(1)): " + str(lambert_W_arg) + " > " + str(-1.0/np.exp(1.0)))
+            print ("gamma: ", gamma)
+            print ("Nx: ", self.Nx)
+            print ("w_rest: ", w_rest)
+            print ("target_weight: ", target_weight)
+            print ("phi: ", phi)
+            print ("!!!!!!!!!!!!")
+        #if self.debug: print "Drawing random number alpha_u: ", alpha_u
```
Comment (setmar) [inline gpu_ocean/SWESimulators/IEWPFOcean.py]: Inactive code should generally be removed.

File: gpu_ocean/SWESimulators/IEWPFOcean.py [hunk 13]
```diff
@@ -0,0 +1,1238 @@
...
+        #oldDebug = self.debug
+        #self.debug = True
+        if self.debug: print ("--------------------------------------")
+        if self.debug: print ("Obtained (lambert_ans k=0, lambert_ans k=-1): ", (lambertw(lambert_W_arg), lambertw(lambert_W_arg, k=-1)))
+        if self.debug: print ("Obtained (alpha k=0, alpha k=-1): ", (alpha_zero, alpha_min1))
+        if self.debug: print ("Checking implicit equation with alpha (k=0, k=-1): ", \
+            (self._implicitEquation(alpha_zero, gamma, self.Nx, a, c), self._implicitEquation(alpha_min1, gamma, self.Nx, a, c)))
+        #if self.debug: print "Chose alpha = ", alpha
```
Comment (setmar) [inline gpu_ocean/SWESimulators/IEWPFOcean.py]: Inactive code should generally be removed.

File: gpu_ocean/SWESimulators/IEWPFOcean.py [hunk 14]
```diff
@@ -0,0 +1,1238 @@
...
+        #if self.debug: print "The implicit equation looked like: \n\t" + \
+        #    "(alpha - 1)"+str(gamma)+" - " + str(self.Nx) + "log(alpha) + " + str(a) + " = 0"
+        #if self.debug: print "Parameters: (gamma, Nx, aj)", (gamma, self.Nx, a)
+
+        alpha = np.sqrt(alpha)
+        if self.debug: print ("alpha = np.sqrt(alpha) ->  ", alpha)
+        if self.debug: print( "--------------------------------------")
+        #debug = oldDebug
```
Comment (setmar) [inline gpu_ocean/SWESimulators/IEWPFOcean.py]: Inactive code should generally be removed.

File: gpu_ocean/SWESimulators/OceanNoiseEnsemble.py [hunk 15]
```diff
@@ -44,7 +46,7 @@ class OceanNoiseEnsemble(BaseOceanStateEnsemble.BaseOceanStateEnsemble):
         
     
     def init(self, driftersPerOceanModel=1):
```
Comment (setmar) [inline gpu_ocean/SWESimulators/OceanNoiseEnsemble.py]: Missing module/class description.

File: gpu_ocean/SWESimulators/PlotHelper.py [hunk 16]
```diff
@@ -425,19 +425,20 @@ def plot(self, ensemble):
         drifterPositions = ensemble.observeDrifters()
         trueDrifterPosition = ensemble.observeTrueDrifters()
         
-        self.true_drifters.set_offsets(drifterPositions)
-        self.true_observations.set_offsets(trueDrifterPosition)
+        #self.true_drifters.set_offsets(drifterPositions)
```
Comment (setmar) [inline gpu_ocean/SWESimulators/PlotHelper.py]: Inactive code should generally be removed. (Goes for this whole section.)

File: gpu_ocean/SWESimulators/gpu_kernels/iewpf_kernels.cu [hunk 17]
```diff
@@ -0,0 +1,204 @@
+/*
+These CUDA kernels implements specific parts of the Implicit Equal-Weights 
+Particle Filter (IEWPF) algorithm.
```
Comment (setmar) [inline gpu_ocean/SWESimulators/gpu_kernels/iewpf_kernels.cu]: Reference(s)?

File: gpu_ocean/SWESimulators/gpu_kernels/iewpf_kernels.cu [hunk 18]
```diff
@@ -0,0 +1,204 @@
...
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+//#include "common.cu"
```
Comment (setmar) [inline gpu_ocean/SWESimulators/gpu_kernels/iewpf_kernels.cu]: Please remove :)

File: gpu_ocean/SWESimulators/gpu_kernels/observationKernels.cu [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+
+
+
+/**
+  * Kernel that observes the velocity of the ocean flow in the drifter positions
+  */
+extern "C" {
+__global__ void observeUnderlyingFlow(
+        //Discretization parameters
+        int nx_, int ny_,
```
Comment (setmar) [inline gpu_ocean/SWESimulators/gpu_kernels/observationKernels.cu]: Some strange indentation stuff going on in this file...

File: gpu_ocean/SWESimulators/gpu_kernels/reductions.cu [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#define NUM_THREADS 128
```
Comment (setmar) [inline gpu_ocean/SWESimulators/gpu_kernels/reductions.cu]: A short comment on the choice of 128 would be nice.

**Comment by setmar** [comment]
Unit tests fail on Tesla M2090/Ubuntu/Python 3:

`martinls@vglserver2:/lustre/storeB/users/martinls/src/gpu-ocean-master/gpu_ocean/tests$ python3 particle_filter_tests.py 5
test_S_matrix (dataAssimilation.IEWPFOcean_test.IEWPFOceanTest) ... ok
test_empty_reduction_buffer (dataAssimilation.IEWPFOcean_test.IEWPFOceanTest) ... ok
test_iewpf_attributes (dataAssimilation.IEWPFOcean_test.IEWPFOceanTest) ... ok
test_kalman_gain (dataAssimilation.IEWPFOcean_test.IEWPFOceanTest) ... ERROR
FAIL
test_localSVD_matrix (dataAssimilation.IEWPFOcean_test.IEWPFOceanTest) ... FAIL
test_local_SVD_to_global_CPU_ref_data (dataAssimilation.IEWPFOcean_test.IEWPFOceanTest) ... FAIL
test_local_SVD_to_global_GPU_ref_data (dataAssimilation.IEWPFOcean_test.IEWPFOceanTest) ... FAIL
FAIL
test_observation_operator_CPU_vs_GPU (dataAssimilation.IEWPFOcean_test.IEWPFOceanTest) ... ok
test_reduction (dataAssimilation.IEWPFOcean_test.IEWPFOceanTest) ... ok
test_set_buffer_to_zero (dataAssimilation.IEWPFOcean_test.IEWPFOceanTest) ... ok

======================================================================
ERROR: test_kalman_gain (dataAssimilation.IEWPFOcean_test.IEWPFOceanTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/lustre/storeB/users/martinls/src/gpu-ocean-master/gpu_ocean/tests/dataAssimilation/IEWPFOcean_test.py", line 244, in test_kalman_gain
    returnKalmanGainTerm=True)
  File "../SWESimulators/IEWPFOcean.py", line 1124, in applyKalmanGain_CPU
    total_K_eta += K_eta[1:-1, 1:-1]
ValueError: operands could not be broadcast together with shapes (40,40) (42,42) (40,40) 

======================================================================
FAIL: test_kalman_gain (dataAssimilation.IEWPFOcean_test.IEWPFOceanTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/lustre/storeB/users/martinls/src/gpu-ocean-master/gpu_ocean/tests/dataAssimilation/IEWPFOcean_test.py", line 40, in tearDown
    self.assertEqual(sys.getrefcount(self.gpu_ctx), 2)
AssertionError: 3 != 2

======================================================================
FAIL: test_localSVD_matrix (dataAssimilation.IEWPFOcean_test.IEWPFOceanTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/lustre/storeB/users/martinls/src/gpu-ocean-master/gpu_ocean/tests/dataAssimilation/IEWPFOcean_test.py", line 165, in test_localSVD_matrix
    assert2DListAlmostEqual(self, localSVD_from_GPU.tolist(), localSVD_from_file.tolist(), 7, "S matrix GPU vs file")
  File "/lustre/storeB/users/martinls/src/gpu-ocean-master/gpu_ocean/tests/testUtils.py", line 213, in assert2DListAlmostEqual
    msg = msg + "First differing element " + str((i,j)) + ":\n" + str(a) + "\n" + str(b) + outro)
AssertionError: -0.9984537959098816 != 7.331716651833468e-20 within 7 places : test case 'S matrix GPU vs file' - lists differs: [[0.0, 0.0, 0.0, 0.0... != [[0.0, 0.0, 0.0, 0.0...

First differing element (1, 13):
-0.9984537959098816
7.331716651833468e-20

======================================================================
FAIL: test_local_SVD_to_global_CPU_ref_data (dataAssimilation.IEWPFOcean_test.IEWPFOceanTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/lustre/storeB/users/martinls/src/gpu-ocean-master/gpu_ocean/tests/dataAssimilation/IEWPFOcean_test.py", line 175, in test_local_SVD_to_global_CPU_ref_data
    assert2DListAlmostEqual(self, test_data.tolist(), results_from_file.tolist(), 10, "test_local_SVD_to_global_CPU")
  File "/lustre/storeB/users/martinls/src/gpu-ocean-master/gpu_ocean/tests/testUtils.py", line 213, in assert2DListAlmostEqual
    msg = msg + "First differing element " + str((i,j)) + ":\n" + str(a) + "\n" + str(b) + outro)
AssertionError: 0.9977135114430458 != 1.0259996349837035 within 10 places : test case 'test_local_SVD_to_global_CPU' - lists differs: [[0.0049999791666927... != [[0.0049999791666927...

First differing element (27, 28):
0.9977135114430458
1.0259996349837035

======================================================================
FAIL: test_local_SVD_to_global_GPU_ref_data (dataAssimilation.IEWPFOcean_test.IEWPFOceanTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/lustre/storeB/users/martinls/src/gpu-ocean-master/gpu_ocean/tests/dataAssimilation/IEWPFOcean_test.py", line 197, in test_local_SVD_to_global_GPU_ref_data
    self.assertAlmostEqual(rel_norm_results, 0.0, places=6)
AssertionError: 4.2217022735668195 != 0.0 within 6 places

======================================================================
FAIL: test_local_SVD_to_global_GPU_ref_data (dataAssimilation.IEWPFOcean_test.IEWPFOceanTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/lustre/storeB/users/martinls/src/gpu-ocean-master/gpu_ocean/tests/dataAssimilation/IEWPFOcean_test.py", line 40, in tearDown
    self.assertEqual(sys.getrefcount(self.gpu_ctx), 2)
AssertionError: 3 != 2

----------------------------------------------------------------------
Ran 10 tests in 23.474s

FAILED (failures=5, errors=1)
Exception ignored in: <bound method CUDAContext.__del__ of <SWESimulators.Common.CUDAContext object at 0x7fb2042b07f0>>
Traceback (most recent call last):
  File "../SWESimulators/Common.py", line 126, in __del__
  File "/usr/lib/python3.5/logging/__init__.py", line 1278, in info
  File "/usr/lib/python3.5/logging/__init__.py", line 1519, in isEnabledFor
TypeError: unorderable types: int() >= NoneType()
Exception ignored in: <bound method CUDAContext.__del__ of <SWESimulators.Common.CUDAContext object at 0x7fb1edaa7a90>>
Traceback (most recent call last):
  File "../SWESimulators/Common.py", line 126, in __del__
  File "/usr/lib/python3.5/logging/__init__.py", line 1278, in info
  File "/usr/lib/python3.5/logging/__init__.py", line 1519, in isEnabledFor
TypeError: unorderable types: int() >= NoneType()
-------------------------------------------------------------------
PyCUDA ERROR: The context stack was not empty upon module cleanup.
-------------------------------------------------------------------
A context was still active when the context stack was being
cleaned up. At this point in our execution, CUDA may already
have been deinitialized, so there is no way we can finish
cleanly. The program will be aborted now.
Use Context.pop() to avoid this problem.
-------------------------------------------------------------------
Aborted (core dumped)`

File: gpu_ocean/SWESimulators/BaseOceanStateEnsemble.py [hunk 21]
```diff
@@ -237,17 +270,47 @@ def observeParticles(self):
             # trueState = [[x1, y1, u1, v1], ..., [xD, yD, uD, vD]]
 
             for p in range(self.numParticles):
-                # Downloading ocean state without ghost cells
-                Hi = self.particles[0].downloadBathymetry()[1]
```
[the remark this answers is not in the document]
Comment (havahol) [reply gpu_ocean/SWESimulators/BaseOceanStateEnsemble.py]: I think then that the default behavior should be to download without ghost cells. Lately I have only been interested in the ghost cells in a few special cases (and not the other way around). 

Will not do anything with this case specifically.

File: gpu_ocean/SWESimulators/IEWPFOcean.py [hunk 22]
```diff
@@ -0,0 +1,1238 @@
...
+            #ensemble.particles[p].drifters.setDrifterPositions(newPos)
+            #print "IEWPF done for particle: ", p
+        # save plot after
+        if infoPlots is not None:
+            self._keepPlot(ensemble, infoPlots, it, 3)
+    
+    ###------------------------    
+    ### GPU Methods
+    ###------------------------
+    # Functions needed for the GPU implementation of IEWPF
```
[the remark this answers is not in the document]
Comment (havahol) [reply gpu_ocean/SWESimulators/IEWPFOcean.py]: All functions used by the iewpf-function could in principle be used from outside the class as well.

File: gpu_ocean/SWESimulators/IEWPFOcean.py [hunk 23]
```diff
@@ -0,0 +1,1238 @@
...
+        x_a = x_a + alpha*xi
+        """
+        if self.debug:
+            print ("gamma: ", gamma)
+            print ("Nx: ", self.Nx)
+            print ("w_rest: ", w_rest)
+            print ("target_weight: ", target_weight)
+            print ("phi: ", phi)
+
+        # 6) Find c
```
[the remark this answers is not in the document]
Comment (havahol) [reply gpu_ocean/SWESimulators/IEWPFOcean.py]: The steps refer to my handwritten notes. I prefer to keep them in the code... :P

File: gpu_ocean/SWESimulators/IEWPFOcean.py [hunk 24]
```diff
@@ -0,0 +1,1238 @@
...
+            # TODO
+            #ensemble.particles[p].drifters.setDrifterPositions(newPos)
+            #print "IEWPF (CPU) done for particle: ", p
+        # save plot after
+        if infoPlots is not None:
+            self._keepPlot(ensemble, infoPlots, it, 3)
+    
+    
+    def iewpf_traditional_CPU(self, ensemble, infoPlots=None, it=None):
+        """
```
[the remark this answers is not in the document]
Comment (havahol) [reply gpu_ocean/SWESimulators/IEWPFOcean.py]: Name changed. Refers to the order of the observations. Both formulations are kept purely for debugging purposes

**Comment by havahol** [comment]
Please see the above commit (or here: https://github.com/havahol/gpu-ocean/commit/1c4c7c2fcdbfc7be343ad1d12fcb9158717fa4dd) for how most of the comments have been resolved. 

About the unit tests: The Kalman gain test was known to fail. The others were a surprise. I had some issues with the tests on my laptop, even though they all pass on my desktop. I will investigate this and resolve it before the next PR (which also will include the commit mentioned above).

**Pull request merged**